File: 1. 단계별로_풀어보기/20~29/27. 그래프와 순회/10. 1697.py
from collections import deque
import sys

input = sys.stdin.readline
output = sys.stdout.write

N, K = map(int, input().split())

resultArr = [None] * 200002
resultArr[N] = 0

visitQueue = deque()
visitQueue.append(N)

while len(visitQueue) > 0:
  now = visitQueue.popleft()
  nowDist = resultArr[now]

  if now > 0 and resultArr[now - 1] == None:
    resultArr[now - 1] = nowDist + 1
    visitQueue.append(now - 1)
  if now < 200000 and resultArr[now + 1] == None:
    resultArr[now + 1] = nowDist + 1
    visitQueue.append(now + 1)
  if now < 100000 and resultArr[2 * now] == None:
    resultArr[2 * now] = nowDist + 1
    visitQueue.append(2 * now)

output(f"{resultArr[K]}")

File: 1. 단계별로_풀어보기/20~29/27. 그래프와 순회/10. 1697.js
// 입력
const fs = require('fs')
const inputArr = fs.readFileSync(0).toString().trim().split('\n')

// 입력 파싱
const [N, K] = inputArr[0].trim().split(' ').map(Number)

// 최단거리를 저장할 배열
const resultArr = Array.from({length: 200002}).fill(null)

resultArr[N] = 0

// 다음번 방문할 숫자들을 저장할 Queue 역할
const visitQueue = Array.from({length: 200002}).fill(null)
let nowIdx = 0
let insertIdx = 0

// 시작지점 Queue 에 삽입
visitQueue[insertIdx++] = N

// 비재귀 BFS 연산
while (nowIdx < insertIdx) {
  const now = visitQueue[nowIdx++]
  const nowDist = resultArr[now]

  if (now > 0 && resultArr[now - 1] === null) {
    resultArr[now - 1] = nowDist + 1
    visitQueue[insertIdx++] = now - 1
  }
  if (now < 200000 && resultArr[now + 1] === null) {
    resultArr[now + 1] = nowDist + 1
    visitQueue[insertIdx++] = now + 1
  }
  if (now < 100001 && resultArr[2 * now] === null) {
    resultArr[2 * now] = nowDist + 1
    visitQueue[insertIdx++] = 2 * now
  }
}

process.stdout.write(`${resultArr[K]}`)
